### PR TITLE
#163426633 Users follow each other

### PR DESCRIPTION
**Pull request by ODINAKACHUKWU**
#### What does this PR do?
Implement users should be able to follow each other

#### Description of Task to be completed?
- write tests for the feature
- users follow each other
- users unfollow each other
- users see authors followed
- users see followers

#### How should this be manually tested?
- clone the repo
- cd into the directory
- cd into the folder and run `npm install` to install dependencies
- run `npm test`
- run `npm run start:dev` to test with Postman

#### What are the relevant pivotal tracker stories?
[#163426633](https://www.pivotaltracker.com/story/show/163426633)

### CODE REVIEW

File: server/controllers/userFollow.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      return httpResponse(res, {
+        statusCode: 201,
+        success: true,
+        message: "You are now following the author"
+      });
+    }
+    return httpResponse(res, {
+      statusCode: 404,
+      success: false,
+      message: "Sorry, you are following this author"
```
Comment (eokwukwe) [inline server/controllers/userFollow.js]: ` "Sorry, you are following this author"` can be changed to ` "Sorry, you are already following this author"` to make it clear that the action is a **_repetition_** of what has been done already.

File: server/controllers/userFollow.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+export const unfollowUser = async (req, res) => {
+  const { userId } = req.user;
+  const authorId = req.userDetails.id;
+  try {
+    if (userId === authorId) {
+      return httpResponse(res, {
+        statusCode: 403,
+        success: false,
+        message: "You cannot follow yourself"
+      });
```
Comment (eokwukwe) [inline server/controllers/userFollow.js]: This is **unfollowUser**. So the message is not correct and the `if` block may not be necessary

**Review by eokwukwe — COMMENTED** [review]
Please take a look at my comments

**Review by eokwukwe — APPROVED** [review]

File: server/controllers/userFollow.js (hunk 2 above)
Comment (ODINAKACHUKWU) [reply server/controllers/userFollow.js]: It has been implemented

File: server/controllers/userFollow.js (hunk 1 above)
Comment (ODINAKACHUKWU) [reply server/controllers/userFollow.js]: It has been implemented

File: server/controllers/userFollow.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+    const follower = await Followers.findOne({
+      where: { authorId, followerId: userId }
+    });
+    if (isEmpty(follower)) {
```
Comment (Ndifreke) [inline server/controllers/userFollow.js]: You can skip isEmpty here and simple test follower. As its value will be null if it did not match any column

**Review by Ndifreke — APPROVED** [review]

File: server/controllers/userFollow.js (hunk 3 above)
Comment (ODINAKACHUKWU) [reply server/controllers/userFollow.js]: Implemented!

File: server/controllers/userFollow.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        statusCode: 403,
+        success: false,
+        message: "You cannot follow yourself"
+      });
+    }
+    const follow = await Followers.findOrCreate({
+      where: { authorId, followerId: userId }
+    });
+    const following = follow[1];
+    if (following === true) {
```
Comment (benfluleck) [inline server/controllers/userFollow.js]: You don't need this true here you can say if(following)

**Review by benfluleck — DISMISSED** [review]

File: server/controllers/userFollow.js (hunk 4 above)
Comment (ODINAKACHUKWU) [reply server/controllers/userFollow.js]: Implemented!

File: server/tests/controller/userFollow/follow.spec.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        .get(`/api/v1/profiles/follow`)
+        .set({ Authorization: token });
+      expect(res.statusCode).to.equal(200);
+      expect(res.body.success).to.be.true;
+      expect(res.body.message).to.equal("Followed authors retrieved");
+      expect(res.body.count).to.equal(2);
+      expect(res.body.data).to.be.an("array");
+    });
+
+    it("Should return you are not following any author if none is followed", async () => {
```
Comment (benfluleck) [inline server/tests/controller/userFollow/follow.spec.js]: `should return an error .....` This test here is not descriptive enough

File: server/tests/controller/userFollow/follow.spec.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        .get(`/api/v1/profiles/follow`)
+        .set({ Authorization: token });
+      expect(res.statusCode).to.equal(404);
+      expect(res.body.success).to.be.false;
+      expect(res.body.message).to.equal("You are not following any author");
+    });
+  });
+
+  describe("GET </api/v1/profiles/follower>", () => {
+    it("User should be able to see all followers", async () => {
```
Comment (benfluleck) [inline server/tests/controller/userFollow/follow.spec.js]: Please follow correct test writing conventions by starting with `should`

File: server/tests/controller/userFollow/follow.spec.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        .get(`/api/v1/profiles/follower`)
+        .set({ Authorization: token });
+      expect(res.statusCode).to.equal(200);
+      expect(res.body.success).to.be.true;
+      expect(res.body.message).to.equal("Followers retrieved");
+      expect(res.body.count).to.equal(2);
+      expect(res.body.data).to.be.an("array");
+    });
+
+    it("Should return you do not have followers if none", async () => {
```
Comment (benfluleck) [inline server/tests/controller/userFollow/follow.spec.js]: Please this test is not descriptive enough

File: server/tests/controller/userFollow/follow.spec.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      expect(res.statusCode).to.equal(200);
+      expect(res.body.success).to.be.true;
+      expect(res.body.message).to.equal(
+        "You are no longer following John Drew"
+      );
+    });
+  });
+
+  describe("GET </api/v1/profiles/follow>", () => {
+    it("User should be able to see all authors followed", async () => {
```
Comment (benfluleck) [inline server/tests/controller/userFollow/follow.spec.js]: Please use correct test conventions please

File: server/tests/controller/userFollow/follow.spec.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      const res = await chai
+        .request(app)
+        .post(`/api/v1/profiles/${followerId}/follow`)
+        .set({ Authorization: token });
+      expect(res.statusCode).to.equal(403);
+      expect(res.body.success).to.be.false;
+      expect(res.body.message).to.equal("You cannot follow yourself");
+    });
+
+    it("User should be able to unfollow an author", async () => {
```
Comment (benfluleck) [inline server/tests/controller/userFollow/follow.spec.js]: Please use correct conventions and start your it block with should

File: server/tests/controller/userFollow/follow.spec.js [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+      const res = await chai
+        .request(app)
+        .post(`/api/v1/profiles/${authorId}/follow`)
+        .set({ Authorization: token });
+      expect(res.statusCode).to.equal(201);
+      expect(res.body.success).to.be.true;
+      expect(res.body.message).to.equal("You are now following John Drew");
+    });
+
+    it("User should not be able to follow self", async () => {
```
Comment (benfluleck) [inline server/tests/controller/userFollow/follow.spec.js]: Please use correct conventions and start your it block with should

File: server/tests/controller/userFollow/follow.spec.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+const jwtConfigs = getJWTConfigs({ option: "authentication" });
+
+chai.use(chaiHttp);
+const { Followers } = models;
+
+beforeEach(async () => {
+  await models.sequelize.sync({ force: true });
+});
+
+describe("Test for users following each other </api/v1/profiles>", () => {
```
Comment (benfluleck) [inline server/tests/controller/userFollow/follow.spec.js]: Please go and look at other tests and change this appropriately

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/routes/index.js [hunk 12]
```diff
@@ -1,8 +1,10 @@
+/* eslint-disable import/no-unresolved */
```
Comment (benfluleck) [inline server/routes/index.js]: Please remove this and edit the eslint config to implement this

File: server/middlewares/findUser/findUser.js [hunk 13]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/no-unresolved */
```
Comment (benfluleck) [inline server/middlewares/findUser/findUser.js]: Please remove this and edit the eslint config

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/middlewares/findUser/findUser.js (hunk 13 above)
Comment (ODINAKACHUKWU) [reply server/middlewares/findUser/findUser.js]: It has been implemented

File: server/routes/index.js (hunk 12 above)
Comment (ODINAKACHUKWU) [reply server/routes/index.js]: It has been implemented

**Review by benfluleck — APPROVED** [review]

**Pull request merged**